File: 1721.swapping-nodes-in-a-linked-list.py
#
# @lc app=leetcode id=1721 lang=python3
#
# [1721] Swapping Nodes in a Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import Optional


class Solution:
    '''
    Find length of the whole list first
    Then locate nodes k and n - k
    '''
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        n = 0
        temp = head
        while temp:
            n += 1
            temp = temp.next
        i, j = None, None
        temp = head
        count =  0
        while temp:
            count += 1
            if count == k:
                i = temp
            if count == n + 1 - k:
                j = temp
            temp = temp.next
        temp = i.val
        i.val = j.val
        j.val = temp
        return head
# @lc code=end

File: 1721.swapping-nodes-in-a-linked-list.cs
/*
 * @lc app=leetcode id=1721 lang=csharp
 *
 * [1721] Swapping Nodes in a Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution 
{
    public ListNode SwapNodes(ListNode head, int k) 
    {
        int n = 0;
        ListNode temp = head;
        while (temp != null)
        {
            n++;
            temp = temp.next;
        }
        ListNode i = null;
        ListNode j = null;
        temp = head;
        int count = 0;
        while (temp != null)
        {
            count++;
            if (count == k)
            {
                i = temp;
            }
            if (count == n + 1 - k)
            {
                j = temp;
            }
            temp = temp.next;
        }
        int val = i.val;
        i.val = j.val;
        j.val = val;
        return head;
    }
}
// @lc code=end

